src/NovelAudioService.cpp

```diff
@@ -0,0 +1,367 @@
+  }
+  else
+  {
+    if (Mix_Playing(_channelMap[soundName]) == 1)
+    {
+      std::cout << "Already playing on channel " << _channelMap[soundName] << std::endl;
+      return;
+    }
+    Mix_PlayChannel(_channelMap[soundName], _sounds[soundName], loops-1);
+  }
+
+}
+
+void NovelAudioService::stopSound(std::string soundName) {
+  Mix_HaltChannel(_channelMap[soundName]);
+}
+
+int NovelAudioService::getSoundChannel(std::string soundName) {
+  auto existingSound = _channelMap.find(soundName);
+  if (existingSound == _channelMap.end())
+  {
+    std::cerr << "ERROR: Cannot get channel for sound " << soundName << std::endl;
+    return -9;
+  }
+  return _channelMap[soundName];
+}
+
+void NovelAudioService::setSoundVolume(std::string soundName, float value) {
+  Mix_VolumeChunk(_sounds[soundName], convertToMixVolume(value));
+}
+
+void NovelAudioService::setSoundPosition(std::string soundName, int angle, int distance) {
+  if (_channelMap.find(soundName) != _channelMap.end() && _channelMap[soundName] != -1)
+  {
+    Mix_SetPosition(_channelMap[soundName], angle, distance);
+  }
+  else
+  {
+    std::cerr << soundName << " is not allocated to a channel already!" << std::endl;
+  }
+}
+
+void NovelAudioService::setSoundDistance(std::string soundName, int distance) {
+    if (_channelMap.find(soundName) != _channelMap.end() && _channelMap[soundName] != -1)
+  {
+    Mix_SetDistance(_channelMap[soundName], distance);
+  }
+  else
+  {
+    std::cerr << soundName << " is not allocated to a channel already!" << std::endl;
+  }
+}
+
+void NovelAudioService::setSoundPanning(std::string soundName, int leftChannelVolume, int rightChannelVolume) {
+    if (_channelMap.find(soundName) != _channelMap.end() && _channelMap[soundName] != -1)
+  {
+    Mix_SetPanning(_channelMap[soundName], leftChannelVolume, rightChannelVolume);
+  }
+  else
+  {
+    std::cerr << soundName << " is not allocated to a channel already!" << std::endl;
+  }
+}
+
+void NovelAudioService::loadMusic(std::string input) {
+  auto exists = _music.find(input);
+  if (exists != _music.end()) return;
+
+  Mix_Music* newMusic = Mix_LoadMUS(input.c_str());
+  if (newMusic != nullptr)
+  {
+    _music[input] = newMusic;
+  }
+  else
+  {
+    std::cerr << Mix_GetError() << std::endl;
+  }
+}
+
+void NovelAudioService::unloadMusic(std::string input) {
+  auto existingMusic = _music.find(input);
+  if (existingMusic == _music.end()) return;
+
+  Mix_FreeMusic(_music[input]);
+  _music.erase(existingMusic);
+}
+
+void NovelAudioService::playMusic() {
+  if (Mix_PausedMusic()) Mix_ResumeMusic();
+}
+
+void NovelAudioService::playMusic(std::string musicName) {
+  auto existingMusic = _music.find(musicName);
+  if (existingMusic == _music.end())
+  {
+    loadMusic(musicName);
+    existingMusic = _music.find(musicName);
+    if (existingMusic == _music.end()) return;
+  }
+  _channelMap[musicName] = -3;
+  Mix_PlayMusic(_music[musicName],-1);
+  _musicTime = SDL_GetPerformanceCounter();
```

Comment:
Nothing wrong with this per se but I can't actually understand why were assigning the SDL performance counter. Could you elaborate?

Reply:
We use the SDL perf counter as a means of getting the time that music started, and then once again when music is paused or halted. This is for the functionality of resuming where you left off (once MP3/OGG support is working; all other formats do not support this and will never resume if you use a fade effect, it will only resume if you pause).

Also, SDL2-Mixer provides no public functionality for keeping track of position

Reply:
Seems like it would be better to define a `Timestamp` type which abstracts this away from the perf counters.